foundationdb/src/future.rs

```diff
@@ -256,226 +255,97 @@ impl AsRef<CStr> for FdbAddress {
-        self.nth_back(0)
-    }
-
-    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
-        if n < self.len() {
-            self.len -= 1 + n as i32;
-            // safe because len < original len
-            let keyvalue = unsafe { self.keyvalues.add(self.len as usize) };
-            Some(FdbValue {
-                _f: self.f.clone(),
-                keyvalue,
-            })
-        } else {
-            self.pos = self.len;
-            None
-        }
-    }
-}
-
 /// A keyvalue you can own
-///
-/// Until dropped, this might prevent multiple key/values from beeing freed.
-/// (i.e. the future that own the data is dropped once all data it provided is dropped)
-pub struct FdbValue {
-    _f: Arc<FdbFutureHandle>,
-    keyvalue: *const fdb_sys::FDBKeyValue,
-}
-
-unsafe impl Send for FdbValue {}
-
-impl Deref for FdbValue {
-    type Target = FdbKeyValue;
-    fn deref(&self) -> &Self::Target {
-        assert_eq_size!(FdbKeyValue, fdb_sys::FDBKeyValue);
-        assert_eq_align!(FdbKeyValue, fdb_sys::FDBKeyValue);
-        unsafe { &*(read_unaligned_struct(self.keyvalue as *const FdbKeyValue)) }
-    }
-}
-impl AsRef<FdbKeyValue> for FdbValue {
-    fn as_ref(&self) -> &FdbKeyValue {
-        self.deref()
-    }
-}
-impl PartialEq for FdbValue {
-    fn eq(&self, other: &Self) -> bool {
-        self.deref() == other.deref()
-    }
-}
-impl Eq for FdbValue {}
-impl fmt::Debug for FdbValue {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        self.deref().fmt(f)
-    }
+#[derive(PartialEq, Eq, Debug)]
+pub struct FdbValue<'a> {
+    key: &'a [u8],
+    value: &'a [u8],
 }
 
-/// A keyvalue owned by a foundationDB future
-///
-/// Because the data it represent is owned by the future in FdbValues, you
-/// can never own a FdbKeyValue directly, you can only have references to it.
-/// This way, you can never obtain a lifetime greater than the lifetime of the
-/// slice that gave you access to it.
-#[repr(transparent)]
-pub struct FdbKeyValue(fdb_sys::FDBKeyValue);
-
-impl FdbKeyValue {
-    /// key
+impl FdbValue<'_> {
     pub fn key(&self) -> &[u8] {
-        unsafe { std::slice::from_raw_parts(self.0.key as *const u8, self.0.key_length as usize) }
+        self.key
     }
 
-    /// value
     pub fn value(&self) -> &[u8] {
-        unsafe {
-            std::slice::from_raw_parts(self.0.value as *const u8, self.0.value_length as usize)
-        }
+        self.value
     }
 }
 
-impl PartialEq for FdbKeyValue {
-    fn eq(&self, other: &Self) -> bool {
-        (self.key(), self.value()) == (other.key(), other.value())
-    }
-}
-impl Eq for FdbKeyValue {}
-impl fmt::Debug for FdbKeyValue {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(
-            f,
-            "({:?}, {:?})",
-            crate::tuple::Bytes::from(self.key()),
-            crate::tuple::Bytes::from(self.value())
-        )
+impl From<fdb_sys::FDBKeyValue> for FdbValue<'_> {
```

Comment:
Not sure about this. I think it disassociates the lifetime of the raw pointers with the returned object